File: scratchpad/gleam/atomic_array/src/atomic_array_ffi.mjs
import { Ok as OkG, Error as ErrorG } from "./gleam.mjs";
import { ComparisonOutOfBounds, ComparisonFailed } from "./atomic_array.mjs";

export function new_signed(size) {
  const bytes = size * BigInt64Array.BYTES_PER_ELEMENT;
  const buffer = new SharedArrayBuffer(bytes);
  return new BigInt64Array(buffer);
}

export function new_unsigned(size) {
  const bytes = size * BigInt64Array.BYTES_PER_ELEMENT;
  const buffer = new SharedArrayBuffer(bytes);
  return new BigUint64Array(buffer);
}

export function get_or_panic(array, index) {
  return Number(Atomics.load(array, index));
}

export function get(array, index) {
  try {
    return new OkG(Number(Atomics.load(array, index)));
  } catch {
    return new ErrorG(undefined);
  }
}

export function size(array) {
  return array.length;
}

export function set(array, index, value) {
  try {
    Atomics.store(array, index, BigInt(value));
    return new OkG(undefined);
  } catch {
    return new ErrorG(undefined);
  }
}

export function add(array, index, amount) {
  try {
    Atomics.add(array, index, BigInt(amount));
    return new OkG(undefined);
  } catch {
    return new ErrorG(undefined);
  }
}

export function exchange(array, index, value) {
  try {
    const x = Atomics.exchange(array, index, BigInt(value));
    return new OkG(Number(x));
  } catch {
    return new ErrorG(undefined);
  }
}

export function compare_exchange(array, index, expected, value) {
  let previous;
  const e = BigInt(expected);
  const v = BigInt(value);
  try {
    previous = Number(Atomics.compareExchange(array, index, e, v));
  } catch {
    return new ErrorG(new ComparisonOutOfBounds());
  }
  if (previous === expected) {
    return new OkG(undefined);
  }
  return new ErrorG(new ComparisonFailed(previous));
}

File: scratchpad/gleam/atomic_array/src/atomic_array_ffi.erl
-module(atomic_array_ffi).
-export([
    new_signed/1, new_unsigned/1, get/2, get_or_panic/2, size/1, set/3, add/3,
    exchange/3, compare_exchange/4
]).

new_signed(Size) ->
    atomics:new(Size, [{signed, true}]).

new_unsigned(Size) ->
    atomics:new(Size, [{signed, false}]).

get_or_panic(Array, Index) ->
    atomics:get(Array, Index + 1).

get(Array, Index) ->
    try
        {ok, atomics:get(Array, Index + 1)}
    catch
        error:badarg -> {error, nil}
    end.

size(Array) ->
    #{size := Size} = atomics:info(Array),
    Size.

set(Array, Index, Value) ->
    try
        atomics:put(Array, Index + 1, Value),
        {ok, nil}
    catch
        error:badarg -> {error, nil}
    end.

add(Array, Index, Amount) ->
    try
        atomics:add(Array, Index + 1, Amount),
        {ok, nil}
    catch
        error:badarg -> {error, nil}
    end.

exchange(Array, Index, Value) ->
    try
        {ok, atomics:exchange(Array, Index + 1, Value)}
    catch
        error:badarg -> {error, nil}
    end.

compare_exchange(Array, Index, Expected, Value) ->
    try
        case atomics:compare_exchange(Array, Index + 1, Expected, Value) of
          ok -> {ok, nil};
          Actual -> {error, {comparison_failed, Actual}}
        end
    catch
        error:badarg -> {error, comparison_out_of_bounds}
    end.
